src/middlewared/middlewared/plugins/smb.py

```diff
@@ -72,6 +72,125 @@ def key_cp(encoding):
             ] + initial
         }
 
+    @private
+    async def validate_admin_groups(self, sid):
+        """
+        Check if group mapping already exists because 'net groupmap addmem' will fail
+        if the mapping exists. Remove any entries that should not be present. Extra
+        entries here can pose a significant security risk. The only default entry will
+        have a RID value of "512" (Domain Admins).
+        In LDAP environments, members of S-1-5-32-544 cannot be removed without impacting
+        entire LDAP environment because this alias exists on the remote LDAP server.
+        """
+        sid_is_present = False
+        ldap = await self.middleware.call('datastore.config', 'directoryservice.ldap')
+        if ldap['ldap_enable']:
+            self.logger.debug("LDAP is enabled. Not removing extra alias entries.")
+            return True
+        proc = await Popen(
+            ['/usr/local/bin/net', 'groupmap', 'listmem', 'S-1-5-32-544'],
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        member_list = (await proc.communicate())[0].decode()
+        if not member_list:
+            return True
+
+        for group in member_list.splitlines():
+            group = group.strip()
+            if group == sid:
+                self.logger.debug(f"SID [{sid}] is already a member of BUILTIN\\administrators")
+                sid_is_present = True
+            if group.rsplit('-', 1)[-1] != "512" and group != sid:
+                self.logger.debug(f"Removing {group} from local admins group.")
+                rem = await Popen(
+                    ['/usr/local/bin/net', 'groupmap', 'delmem', 'S-1-5-32-544', group],
+                    stdout=subprocess.PIPE, stderr=subprocess.PIPE
+                )
+                remout = await rem.communicate()
+                if rem.returncode != 0:
+                    raise CallError(f'Failed to remove sid [{sid}] from S-1-5-32-544: {remout[1].decode()}')
+
+        if sid_is_present:
+            return False
+        else:
+            return True
+
+    @private
+    async def wbinfo_gidtosid(self, gid):
+        verrors = ValidationErrors()
+        proc = await Popen(
+            ['/usr/local/bin/wbinfo', '--gid-to-sid', f"{gid}"],
+            stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        output = await proc.communicate()
+        if proc.returncode != 0:
+            if "WBC_ERR_WINBIND_NOT_AVAILABLE" in output[1].decode():
+                return "WBC_ERR_WINBIND_NOT_AVAILABLE"
```

Comment:
It is good enough now, but if you are interested, the proper way to define constant is

`WBC_ERR_WINBIND_NOT_AVAILABLE = object()`

or

```
class WBC_ERR(enum.Enum):
    WINBIND_NOT_AVAILABLE = 1
```